compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirContractChecker.kt

```diff
@@ -0,0 +1,46 @@
+/*
+ * Copyright 2010-2021 JetBrains s.r.o. and Kotlin Programming Language contributors.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
+ */
+
+package org.jetbrains.kotlin.fir.analysis.checkers.declaration
+
+import org.jetbrains.kotlin.fir.FirFakeSourceElementKind
+import org.jetbrains.kotlin.fir.analysis.checkers.context.CheckerContext
+import org.jetbrains.kotlin.fir.analysis.diagnostics.DiagnosticReporter
+import org.jetbrains.kotlin.fir.analysis.diagnostics.FirErrors
+import org.jetbrains.kotlin.fir.contracts.FirResolvedContractDescription
+import org.jetbrains.kotlin.fir.declarations.FirContractDescriptionOwner
+import org.jetbrains.kotlin.fir.declarations.FirFunction
+import org.jetbrains.kotlin.fir.expressions.*
+
+object FirContractChecker : FirFunctionChecker() {
+    private const val UNEXPECTED_CONSTRUCTION = "unexpected construction in contract description"
+
+    override fun check(declaration: FirFunction<*>, context: CheckerContext, reporter: DiagnosticReporter) {
+        if (declaration !is FirContractDescriptionOwner ||
+            declaration.contractDescription !is FirResolvedContractDescription
+        ) {
+            return
+        }
+
+        // Any statements that [ConeEffectExtractor] cannot extract effects will be in `unresolvedEffects`.
+        for (statement in (declaration.contractDescription as FirResolvedContractDescription).unresolvedEffects) {
+            if (statement.source == null || statement.source!!.kind is FirFakeSourceElementKind) continue
+
+            if (!statement.isAllowedInContract) {
+                reporter.report(FirErrors.ERROR_IN_CONTRACT_DESCRIPTION.on(statement.source!!, UNEXPECTED_CONSTRUCTION), context)
+            }
+        }
+    }
+
+    // See what visit methods are overridden at [ConeEffectExtractor]
+    private val FirStatement.isAllowedInContract: Boolean
```

Comment:
I think this is unnecessary. If some construction is legal, but it's not a effect description then diagnostic should be reported on it. Consider next example:
```
fun foo(x: Any) contract [
    x is String
] {
    ...
}
```
Such contract is meaningless, but you don't report anything on it (as I can see by code of checker)

Reply:
Right, removed, and reports all unresolved effects as mentioned earlier.